### ISSUE TITLE
State of this repository

### ISSUE THREAD

**Issue by sl33k** [CTSRD-CHERI/sail-cheri-riscv#16]
Hey,

I wanted to use this sail model for a thesis. 
I've noticed that there is a `TODO` file in the `src/` folder which has some rather cryptic points on what is left to be done.
Therefore my question is, how complete is this sail implementation? Is it in a usable state or are large parts missing?

Thanks and kind regards,

Malte

**Comment by rmn30** [comment]
Hi Malte,

I believe this repo is still used by folks working on CHERI-RISCV at the University of Cambridge and should be fairly complete and up to date. I recently moved on to a new job so I no longer maintain it, although I might be able to answer questions. I'm not sure who the current maintainer is (hopefully they will show up here if they haven't already been in touch). 

Regarding the TODO file that list is a bit old and probably out of date and should probably be removed. From memory:
- AUIPC -- refers to the [AUIPC instruction](https://github.com/CTSRD-CHERI/sail-cheri-riscv/blob/master/src/cheri_insts.sail#L1204) which is now implemented
- parameterization via CLEN -- the model now supports 128-bit and 64-bit capabilities (although most effort is on 128 bit)
- various fixmes -- you might find various XXX and FIXMEs in the code still, hopefully nothing too important.

**Comment by jrtc27** [comment]
I think an important question is: what would you like to use the Sail model for? If you want to do formal reasoning then it's the right thing to use, but if you just want to run compiled code then we have better alternatives like QEMU and Spike ports, as well as hardware implementations in Bluespec SystemVerilog (which are tested against this as a golden model/specification), especially if you want to run an operating system, since the model currently lacks the basic devices needed to do anything useful there.

**Comment by sl33k** [comment]
First of all, thank you for your comprehensive answers!
Regarding the usage of this model: It's to be used in thesis evaluating the effect of security-hardened architectures on hardware faults, which is why I want to use the golden model (is this a correct assumption?) of this architecture. No devices beside simple I/O (HTIF) are required.
Some further questions:

- I see that this model is tested with [TestRig](https://github.com/CTSRD-CHERI/TestRIG), which uses a random instruction generator. Are there any toolchain porting efforts for RISC-V (such as [llvm](https://github.com/CTSRD-CHERI/llvm-project))?

- Could you elaborate on the `CLEN` parametrization? Is only CHERI-compressed supported? And if so, @rmn30 any pointers on how to implement the uncompressed variant? How much effort would such a port incur approximately?

**Comment by jrtc27** [comment]
> First of all, thank you for your comprehensive answers!
> Regarding the usage of this model: It's to be used in thesis evaluating the effect of security-hardened architectures on hardware faults, which is why I want to use the golden model (is this a correct assumption?) of this architecture. No devices beside simple I/O (HTIF) are required.
> Some further questions:
> 
> * I see that this model is tested with [TestRig](https://github.com/CTSRD-CHERI/TestRIG), which uses a random instruction generator. Are there any toolchain porting efforts for RISC-V (such as [llvm](https://github.com/CTSRD-CHERI/llvm-project))?

Our LLVM has a CHERI RISC-V backend that is used to compile a range of software (CheriBSD, BLL, CheriFreeRTOS, CHERI-RTEMS, CHERI-newlib).

> * Could you elaborate on the `CLEN` parametrization? Is only CHERI-compressed supported? And if so, @rmn30 any pointers on how to implement the uncompressed variant? How much effort would such a port incur approximately?

Yes, we only currently supported compressed 64-bit capabilities and compressed 128-bit capabilities. For uncompressed 256-bit capabilities you could take https://github.com/CTSRD-CHERI/sail-cheri-mips/blob/master/cheri/cheri_prelude_256.sail and adapt it to match this model (the encoding is the same other than needing a flag bit, but the types and helpers are slightly different between models). The instruction encodings for the various load/store capability variants (including AMOs) would be conceptually "wrong", as they were chosen based on the size of capabilities (on RV32I with 64-bit capabilities we use RV64I's LD/SD/etc, on RV64I with 128-bit capabilities we use RV128I's hypothesised LQ/SQ/etc), so ideally 256-bit capabilities would use instructions encoded to look like LQ2/SQ2/etc, but that's not strictly necessary, things will work, including the LLVM assembler. You won't, however, be able to compile C code, as there is no support in the compiler for 256-bit capabilities on CHERI RISC-V (and it is being dropped from CHERI MIPS).

Is there a reason why you want to operate with the uncompressed capability encoding? Whilst its guarantees of full precision with plenty of metadata bits to spare are attractive, it is far too expensive to be realistically deployed, and also causes issues for software where 16-byte alignment is a common default in the ABI, which is too small for 32-byte capabilities. Our software stacks have already started to drop support for it, and we don't see it coming back.

**Comment by jrtc27** [comment]
> I want to use the golden model (is this a correct assumption?) of this architecture.

In theory, yes, and for most things it is, but currently has a glaring deficiency of not supporting all the atomics with either a capability base or a capability value.

**Comment by rwatson** [comment]
Hi Malte:

Just to answer the first question, do take a look at our recently added CHERI-RISC-V landing page:

	https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/cheri-risc-v.html

Robert

**Comment by sl33k** [comment]
@rwatson Thanks! Can't believe I missed that page. Weirdly enough is does not seem to appear on the general CHERI pages sidebar (located [here](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/)), maybe that's something you'd want to fix?

@jrtc27 Thanks for detailed description. 

>  Is there a reason why you want to operate with the uncompressed capability encoding?

The reason for uncompressed capabilities is that single-event-upsets (e.g. bitflips) would possibly be more easily detectable when fine-grained bounds exists. No C-Compiler support seems like a show stopper though.

>  In theory, yes, and for most things it is, but currently has a glaring deficiency of not supporting all the atomics with either a capability base or a capability value.

Could you clarify further what you mean by that? Are you talking about RISC-V atomics?

**Comment by rwatson** [comment]
> @rwatson Thanks! Can't believe I missed that page. Weirdly enough is does not seem to appear on the general CHERI pages sidebar (located [here](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/)), maybe that's something you'd want to fix?

That’s because when I wrote “recently added,” it was very literally true. :-) We created the page about a month ago, but only linked it up yesterday. It should now be findable with Google, and linked from the CHERI page. If there’s anywhere else we should link it, do let us know.

@jrtc27 is referring to missing specifications for capability-extended AMOs in the CHERI-RISC-V Sail description. In all other senses, Sail is the gold model against which we are doing tandem verification for our sundry implementations.
